consensus/state_test.go

```diff
@@ -291,6 +294,51 @@ func TestStateFullRoundNil(t *testing.T) {
 	validatePrevoteAndPrecommit(t, cs, round, 0, vss[0], nil, nil)
 }
 
+// Tests that consensus state can be saved and reloaded from DB after committing a block.
+func TestStateFullRound1Reload(t *testing.T) {
+	blockDB := dbm.NewMemDB()
+	state, privVals := randGenesisState(1, false, 10)
+	vss := make([]*validatorStub, 1)
+	cs := newConsensusStateWithConfigAndBlockStore(config, state, privVals[0], counter.NewCounterApplication(true), blockDB)
+	vss[0] = NewValidatorStub(privVals[0], 0)
+	incrementHeight(vss[1:]...)
+	height, round := cs.Height, cs.Round
+
+	// NOTE: buffer capacity of 0 ensures we can validate prevote and last commit
+	// before consensus can move to the next height (and cause a race condition)
+	cs.eventBus.Stop()
+	eventBus := types.NewEventBusWithBufferCapacity(0)
+	eventBus.SetLogger(log.TestingLogger().With("module", "events"))
+	cs.SetEventBus(eventBus)
+	eventBus.Start()
+
+	voteCh := subscribe(cs.eventBus, types.EventQueryVote)
+	propCh := subscribe(cs.eventBus, types.EventQueryCompleteProposal)
+	newRoundCh := subscribe(cs.eventBus, types.EventQueryNewRound)
+
+	startTestRound(cs, height, round)
+
+	<-newRoundCh
+
+	// grab proposal
+	re := <-propCh
+	propBlockHash := re.(types.EventDataRoundState).RoundState.(*cstypes.RoundState).ProposalBlock.Hash()
+
+	<-voteCh // wait for prevote
+	validatePrevote(t, cs, round, vss[0], propBlockHash)
+
+	<-voteCh // wait for precommit
+
+	// we're going to roll right into new height
+	<-newRoundCh
```

Comment:
This doesn't actually test that we roll into a new height. (newRoundCh just ensures that we go to a new round).
We should actually test that we go to a new height... and I'm not sure that we need the other subscription tests above (since going to a new height pre-supposes prevotes and precommits, and this test is testing reloading... all this setup is making the point of the test less obvious.

Reply:
Yeah, this test was basically just copied from the tests above, with the addition of doing the state reload. I'll read through the stuff you referenced in common_test and rework this to be more straightforward.